src/ir/replacer.h

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright 2021 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef wasm_ir_replacer_h
+#define wasm_ir_replacer_h
+
+#include <wasm-traversal.h>
+#include <wasm.h>
+
+namespace wasm {
+
+// A map of replacements to perform on expressions. After filling the map,
+// simply do a walk(..) and they will be replaced.
+//
+// This is useful for a general replacement of expressions. Some passes can
+// store pointers to expressions and update only those, but then there is the
+// nesting problem,
+//
+//  (foo
+//   (bar
+//    (baz
+//
+// If we replace bar with something then the pointer to baz may change. Doing
```

Comment:
I'm not sure if I understand. If we replace `bar` with something new, which doesn't point `baz`, why does the pointer to `baz` matter?

Reply:
Say we want to replace baz with baz2. Then we need to make the pointer to baz point to baz2.

if we have `bar->baz` (bar points to baz as its single child), and we want to have `bar2->baz2`, then if we replace bar first we end up with `bar2->baz` (a new object bar2 that also points to baz), and then we need to modify bar2's pointer to its child, not bar. (so ifwe saved a pointer to `&bar->baz` , that would be wrong)